src/main/java/com/firebolt/jdbc/connection/FireboltConnectionServiceSecret.java

```diff
@@ -0,0 +1,128 @@
+        this.fireboltAccountIdService = new FireboltAccountIdService(createFireboltAccountRetriever(httpClient, objectMapper, "resolve", FireboltAccount.class));
+        this.fireboltEngineService = new FireboltEngineInformationSchemaService(this);
+        connect();
+    }
+
+    private <T> FireboltAccountRetriever<T> createFireboltAccountRetriever(OkHttpClient httpClient, ObjectMapper objectMapper, String path, Class<T> type) {
+        return new FireboltAccountRetriever<>(httpClient, objectMapper, this, loginProperties.getUserDrivers(), loginProperties.getUserClients(), loginProperties.getHost(), path, type);
+    }
+
+    @Override
+    protected void authenticate() throws SQLException {
+        String account = loginProperties.getAccount();
+        if (account == null) {
+            throw new FireboltException("Cannot connect: account is missing");
+        }
+        if (loginProperties.isSystemEngine() && loginProperties.getDatabase() == null) {
+            throw new FireboltException("The database with the name null could not be found", RESOURCE_NOT_FOUND);
+        }
+        String accessToken = getAccessToken(loginProperties).orElse(StringUtils.EMPTY);
+        sessionProperties = getSessionPropertiesForSystemEngine(accessToken, account);
+        assertDatabaseExisting(loginProperties.getDatabase());
+        if (!loginProperties.isSystemEngine()) {
+            sessionProperties = getSessionPropertiesForNonSystemEngine();
+        }
+    }
+
+    private FireboltProperties getSessionPropertiesForNonSystemEngine() throws SQLException {
+        sessionProperties = sessionProperties.toBuilder().engine(loginProperties.getEngine()).build();
+        Engine engine = fireboltEngineService.getEngine(loginProperties);
+        return loginProperties.toBuilder().host(engine.getEndpoint()).engine(engine.getName()).systemEngine(false).database(engine.getDatabase()).build();
+    }
+
+    @Override
+    protected void assertDatabaseExisting(String database) throws SQLException {
+        if (database != null && !fireboltEngineService.doesDatabaseExist(database)) {
+            throw new FireboltException(format("Database %s does not exist", database), RESOURCE_NOT_FOUND);
+        }
+    }
+
+    private FireboltProperties getSessionPropertiesForSystemEngine(String accessToken, String account) throws FireboltException {
+        String systemEngineEndpoint = fireboltGatewayUrlService.getUrl(accessToken, account);
+        String accountId = fireboltAccountIdService.getValue(accessToken, account);
+        return this.loginProperties
+                .toBuilder()
+                .systemEngine(true)
+                .additionalProperties(Map.of())
+                .compress(false)
+                .accountId(accountId)
+                .host(UrlUtil.createUrl(systemEngineEndpoint).getHost())
+                .build();
```

Comment:
Here, we don't set the database. Do we set it elsewhere?

Reply:
Sure. Here we update initial `loginPorperties` where all initially provided properties are updated (including DB if it has been supplied). The only reason of this code is to override some properties, e.g. systemEngine=true (because this is system engine), corepress=false (because system engine cannot compress data) etc.